### PR TITLE
New Conditional meta game action

### PR DESCRIPTION
**Pull request by jamesoddy**
Expects parameters:

`gameAction:`
`condition?:`
`else:`

Will resolve gameAction: if condition: is true or will resolve else: if condition: is not true.  If condition: is not defined then will resolve else: if gameAction: cannot effect game state.

### CODE REVIEW

File: server/game/GameActions/ConditionalGameAction.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import AbilityContext = require('../AbilityContext');
+
+export interface ConditionalActionProperties extends GameActionProperties {
+    gameAction: GameAction;
+    condition: boolean | ((context: AbilityContext) => boolean);
+    else?: GameAction;
+}
+
+export class ConditionalGameAction extends GameAction {
+    effect = 'do a conditional action';
```
Comment (jeremylarner) [inline server/game/GameActions/ConditionalGameAction.ts]: I think we should probably give this an effect message which tells the player what it's going to do

File: server/game/GameActions/ConditionalGameAction.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    constructor(properties: ConditionalActionProperties | ((context: AbilityContext) => ConditionalActionProperties)) {
+        super(properties);
```
Comment (jeremylarner) [inline server/game/GameActions/ConditionalGameAction.ts]: The constructor here shouldn't be needed?

File: server/game/GameActions/ConditionalGameAction.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import { GameAction, GameActionProperties } from './GameAction';
+import AbilityContext = require('../AbilityContext');
+
+export interface ConditionalActionProperties extends GameActionProperties {
+    gameAction: GameAction;
+    condition: boolean | ((context: AbilityContext) => boolean);
```
Comment (jeremylarner) [inline server/game/GameActions/ConditionalGameAction.ts]: Would this ever be boolean?  I think it should probably always be a function....

File: server/game/GameActions/ConditionalGameAction.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    effect = 'do a conditional action';
+
+    constructor(properties: ConditionalActionProperties | ((context: AbilityContext) => ConditionalActionProperties)) {
+        super(properties);
+    }
+
+    getProperties(context: AbilityContext, additionalProperties = {}): ConditionalActionProperties {
+        let properties = super.getProperties(context, additionalProperties) as ConditionalActionProperties;
+        properties.condition = properties.condition || true;
+        return properties;
```
Comment (jeremylarner) [inline server/game/GameActions/ConditionalGameAction.ts]: This should probably be handled with defaultProperties.  You should put something in here though which sets the two game action default targets to be this actions target

File: server/game/GameActions/ConditionalGameAction.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        if(typeof properties.condition === 'function') {
+            return properties.condition(context);
+        }
+        return properties.condition as boolean;
+    }
+
+    hasLegalTarget(context: AbilityContext, additionalProperties = {}): boolean {
+        let properties = this.getProperties(context, additionalProperties);
+        if(this.meetsCondition(context, properties)) {
+            return properties.gameAction.hasLegalTarget(context, additionalProperties) || properties.else.hasLegalTarget(context, additionalProperties);
```
Comment (jeremylarner) [inline server/game/GameActions/ConditionalGameAction.ts]: This will throw an error if gameAction isn't legal and properties.else isn't defined

**Comment by jeremylarner** [comment]
The only thing I think is potentially confusing here is that if the condition is true but gameAction is illegal, what happens?  I can see cases where you want else to happen, but also cases where you don't....

**Comment by jamesoddy** [comment]
Potential cards that could use this
```
Deathseeker, Mirumoto Raitsugu, Calling in Favors, Karada District
Akodo makoto
CiF and Karada are a bit of a mess atm
so you might need to be a bit creative to get them to work
Also the new dragon holding might be a good fit for a conditional action
```

**Comment by jeremylarner** [comment]
This needs to be PRed to develop (or somewhere else)

**Pull request closed without merging**